### ISSUE TITLE
How to rebuild adam with a specific version of spark and scala ?

### ISSUE THREAD

**Issue by bioinfornatics** [bigdatagenomics/adam#2052]
Dear,

Currently, we use spark `2.1.0` from hortonworks `2.6.0.3-8` with scala `2.11.8`

In our system we use the spark jars provided by hortonworks using a local nexus.

``` bash
$ ls /usr/hdp/current/spark2-client/jars/spark*
/usr/hdp/current/spark2-client/jars/spark-catalyst_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-cloud_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-core_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-graphx_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-hive_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-hive-thriftserver_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-launcher_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-mllib_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-mllib-local_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-network-common_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-network-shuffle_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-repl_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-sketch_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-sql_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-streaming_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-tags_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-unsafe_2.11-2.1.0.2.6.0.3-8.jar
/usr/hdp/current/spark2-client/jars/spark-yarn_2.11-2.1.0.2.6.0.3-8.jar
```

So in order to build adam I do:

```bash
$ git clone https://github.com/bigdatagenomics/adam.git
$ pushd adam
$ git checkout adam-parent-spark2_2.11-0.24.0
$ mvn -Dhadoop.version=2.7.3 \
           -Dspark.version=2.1.0.2.6.0.3-8  \
           -Dscala.version=2.11.8 \
           -Dscala.version.prefix=2.11 \
           -U \
           clean package install 
```

But the build process for adam-core fail with the following stack trace:

```
[INFO] /usr/local/bdg_adam/adam-core/src/main/java:-1: info: compiling
[INFO] /usr/local/bdg_adam/adam-core/src/main/scala:-1: info: compiling
[INFO] /usr/local/bdg_adam/adam-core/target/generated-sources/src/main/scala:-1: info: compiling
[INFO] Compiling 135 source files to /usr/local/bdg_adam/adam-core/target/2.11.8/classes at 1537199877170
[ERROR] error: missing or invalid dependency detected while loading class file 'package.class'.
[INFO] Could not access term annotation in package org.apache.spark,
[INFO] because it (or its dependencies) are missing. Check your build definition for
[INFO] missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
[INFO] A full rebuild may help if 'package.class' was compiled against an incompatible version of org.apache.spark.
[WARNING] warning: Class org.apache.spark.annotation.InterfaceStability not found - continuing with a stub.
[ERROR] error: missing or invalid dependency detected while loading class file 'functions.class'.
[INFO] Could not access term annotation in package org.apache.spark,
[INFO] because it (or its dependencies) are missing. Check your build definition for
[INFO] missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
[INFO] A full rebuild may help if 'functions.class' was compiled against an incompatible version of org.apache.spark.
[ERROR] /usr/local/bdg_adam/adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala:66: error: object json4s is not a member of package org
[ERROR] import org.json4s.JsonDSL._
[ERROR]            ^
[ERROR] /usr/local/bdg_adam/adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala:67: error: object json4s is not a member of package org
[ERROR] import org.json4s.jackson.JsonMethods._
[ERROR]            ^
[ERROR] /usr/local/bdg_adam/adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala:109: error: object json4s is not a member of package org
[ERROR] import org.json4s.jackson.JsonMethods._
[ERROR]            ^
[WARNING] warning: Class org.apache.spark.annotation.InterfaceStability not found - continuing with a stub.
```

So they are a problem with json4s package or scala. I do not know.
If we take a look to the user maven repository, we can see that maven have not donwloaded the json4s jar.

```bash
$ cd ${MAVEN_HOME}
$ ls org/json4s
ls: cannot access org/json4s: No such file or directory
```

As json4s seem to be required I try to force maven to use it . For this purpose I add in `adam-core/pom.xml` the following dependency before rebuilding the project

```
<dependency>
  <groupId>org.json4s</groupId>
  <artifactId>json4s-core_2.11</artifactId>
  <version>3.2.11</version>
</dependency>
```

```bash
$ sed -i '/<dependencies>/a <dependency>\n  <groupId>org.json4s</groupId>\n  <artifactId>json4s-core_2.11</artifactId>\n  <version>3.2.11</version>\n</dependency>' adam-core/pom.xml
```
This step have been done too for `json4s-ast` and `json4s-jackson`

While the json4s jars are now installed in the `${MAVEN_HOME}` directory, I have still the same build fail.

Did you have any idea what is the problem here ?

Thanks a lot, have a nice day

**Comment by bioinfornatics** [comment]
I have the same build problem by using any spark version available on maven repo (`2.3.0` , `2.2.0`, `2.1.0`)

```bash
for spark_version in 2.3.0 2.2.0 2.1.0; do 
  mvn clean package install \
          -e \
          -Dhadoop.version=2.7.3 \
          -Dspark.version=${spark_version} \
          -Dscala.version=2.11.8 \
          -Dscala.version.prefix=2.11 \
          -Dscala-2.11 \
           -U
done
```

**Comment by bioinfornatics** [comment]
I close this issue until i have more information
